src/library_pthread.js

```diff
@@ -1219,18 +1219,40 @@ var LibraryPThread = {
   },
 #endif // MAIN_MODULE
 
-  $checkMailbox__deps: ['$callUserCallback'],
+  $checkMailbox__deps: ['$callUserCallback',
+                        '_emscripten_thread_mailbox_await'],
   $checkMailbox: function() {
     // Only check the mailbox if we have a live pthread runtime. We implement
     // pthread_self to return 0 if there is no live runtime.
-    if (_pthread_self()) {
+    var pthread_ptr = _pthread_self();
+    if (pthread_ptr) {
+      // If we are using Atomics.waitAsync as our notification mechanism, wait for
+      // a notification before processing the mailbox to avoid missing any work.
+      __emscripten_thread_mailbox_await(pthread_ptr);
       callUserCallback(() => __emscripten_check_mailbox());
     }
   },
 
-  _emscripten_notify_mailbox__deps: ['$checkMailbox'],
-  _emscripten_notify_mailbox__sig: 'vppp',
-  _emscripten_notify_mailbox: function(targetThreadId, currThreadId, mainThreadId) {
+  _emscripten_thread_mailbox_await__deps: ['$checkMailbox'],
+  _emscripten_thread_mailbox_await__sig: 'vp',
+  _emscripten_thread_mailbox_await: function(pthread_ptr) {
+    if (typeof Atomics.waitAsync === 'function') {
+      // TODO: How to make this work with wasm64?
+      var wait = Atomics.waitAsync(HEAP32, pthread_ptr >> 2, pthread_ptr);
```

Comment:
So the waiting is don't on the pthread pointer itself?  (not some member?)

Reply:
Yes. It doesn't really matter what address we wait on as long as we know what its value will be when we execute `Atomics.waitAsync`. The first member of the pthread struct is conveniently a pointer to itself, so we know what value it will have when we wait on it.